Custom icons for SSO providers

omarkhan opened the pull request:
**Background:** Only font awesome classes can currently be used to provide icons for third party authentication login/registration buttons. This pull request allows admin users to upload images to use as icons, when a font awesome class is not available for a given provider.

**JIRA ticket:** [RAIN-3](https://openedx.atlassian.net/browse/RAIN-3)

**Screenshots:**

Configuration:
<img width="938" alt="screen shot 2016-03-24 at 15 03 48" src="https://cloud.githubusercontent.com/assets/793379/14011500/bba66c2c-f1d1-11e5-8d4b-6ae28576d0ab.png">

Login form:
<img width="631" alt="screen shot 2016-03-24 at 14 59 12" src="https://cloud.githubusercontent.com/assets/793379/14011454/5911dd94-f1d1-11e5-8481-1c749f6af313.png">

Registration form:
<img width="630" alt="screen shot 2016-03-24 at 14 59 51" src="https://cloud.githubusercontent.com/assets/793379/14011458/5fae66d6-f1d1-11e5-9a23-e63e5d3868d8.png">

Hinted login:
<img width="622" alt="screen shot 2016-03-24 at 15 00 27" src="https://cloud.githubusercontent.com/assets/793379/14011462/6571faa6-f1d1-11e5-8b65-378612e02b46.png">

**Sandbox:** http://pr11871.sandbox.opencraft.com/ (persistent database)

**Merge deadline:** Friday 25 March 2016

**Author notes and concerns:** Untested so far:

- Stanford theme

**Settings**
```yaml
FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
  ENABLE_THIRD_PARTY_AUTH: true
EDXAPP_ENV_EXTRA:
    THIRD_PARTY_AUTH_BACKENDS:
        - social.backends.google.GoogleOAuth2
        - social.backends.linkedin.LinkedinOAuth2
        - social.backends.facebook.FacebookOAuth2
        - social.backends.azuread.AzureADOAuth2
        - third_party_auth.saml.SAMLAuthBackend
        - third_party_auth.lti.LTIAuthBackend
disable_edx_services: true
```

edx-webhook commented:
Thanks for the pull request, @omarkhan! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=11871

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -109,6 +118,12 @@ class Meta(object):
+        """ Ensure that either `icon_class` or `icon_image` is set """
+        super(ProviderConfig, self).clean()
+        if bool(self.icon_class) == bool(self.icon_image):
+            raise ValidationError('Either an icon class or an icon image must be given')
```
Should say, 'Either an icon class or an icon image must be given (but not both)' because specifying both could also trigger this error.

bradenmacdonald commented on common/djangoapps/third_party_auth/models.py:
```diff
@@ -70,12 +70,21 @@ class ProviderConfig(ConfigurationModel):
     )
+    icon_image = models.ImageField(
+        blank=True,
+        help_text=_(
```
So far we have not been translating these admin strings, because they're only seen by DevOps, and it would create extra work for all the translators with little benefit.

bradenmacdonald commented on themes/stanford-style/lms/templates/register-form.html:
```diff
@@ -26,7 +26,14 @@ <h3 class="message-title">${_("The following errors occurred while processing yo
+        % if enabled.icon_class:
+        <span class="icon fa ${enabled.icon_class}" aria-hidden="true"></span>
+        % else:
+        <img class="icon" src="${enabled.icon_image.url}" aria-hidden="true" />
```
`<img` tags should always have an `alt` attribute; since we don't want screen readers to read anything here, we've marked it as aria-hidden but you should still put `alt=""`

bradenmacdonald commented on lms/templates/register-form.html:
```diff
@@ -26,7 +26,14 @@ <h3 class="message-title">${_("The following errors occurred while processing yo
+      % if enabled.icon_class:
+      <span class="icon fa ${enabled.icon_class}" aria-hidden="true"></span>
+      % else:
+      <img class="icon" src="${enabled.icon_image.url}" aria-hidden="true" />
```
`<img` tags should always have an `alt` attribute; since we don't want screen readers to read anything here, we've marked it as aria-hidden but you should still put `alt=""`

bradenmacdonald commented:
@omarkhan Sorry for not realizing this earlier, but we're also going to need a new field to specify the background color of the square which appears behind the image on the combined form. See the approved mockup attached to RAIN-2 or the discovery doc.

bradenmacdonald commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -385,6 +385,10 @@ $sm-btn-linkedin: #0077b5;
             color: $white;
+
+            img {
+                width: 100%;
```
I'm worried this icon will be too big compared to the Google and Facebook Font Awesome icons. When creating mockups for these stories, I think I just used some styles like `"background-image: url(...); background-size: 24px auto; background-position: center;"` on the `.icon` element (something like that).

Can you create a new image that's just a square white PNG, 100px x 100px, and use that for your screenshots? And enable the Google and Facebook providers too? It will give us a better idea of the sizing.

Just specifying the width could also break the layout if some user uploads a logo that has dimensions like 40px x 70px; it would be awesome if we could handle that case, maybe with a max-height rule.

bradenmacdonald commented:
@omarkhan We probably don't need to implement any support for this on the old registration/login pages (only on the combined page). The old one will be removed soon in #11549

omarkhan commented on scripts/reset-test-db.sh:
```diff
@@ -22,6 +22,8 @@
 #
 ############################################################################
 
+set -e
```
I put this in as a safety measure because:

- I had added [the private settings file described here](https://github.com/edx/edx-platform/wiki/Developing-on-the-edX-Developer-Stack#making-the-local-servers-run-faster) to speed up my devstack
- That private settings file overrides the `INSTALLED_APPS` changes in the bok choy settings file, which means that `django_extensions` is no longer in there
- That causes this script to fail, becase the `reset_db` command is no longer available
- But the script would just ignore the failure and carry on as per bash's default behaviour, causing much confusion

bradenmacdonald commented:
@omarkhan 
* Was it necessary to update `common/test/db_cache/` ? Normally we leave that folder alone, and every once in a while someone from testeng or devops will re-create the cache. Its purpose is just to speed things up by including many of the migrations already in place, but not necessarily all.
* The "test icon" has a lot of padding built into the image and I think it misrepresents the sizing of the icons. I think you should create an SVG image that is a white square (with no padding around it) and use that in the tests and sandbox. It should then appear to be centered and about the same size as the Facebook 'f' and Google 'g'
* It seems like you have removed the ability to configure MEDIA_URL from env code like `lms/envs/aws.py`? I don't think that's a reasonable change. If I run some instance and want to serve my media from a separate subdomain, I need to be able to configure that.

omarkhan commented:
@bradenmacdonald:

- My bok choy tests were failing before I updated the db cache, but I'm not sure why. I will try reverting the changes and running them again
- I will update the image as per your suggestion, this was just a placeholder while I tried to get the tests working
- The ability to configure the MEDIA_URL from env code is still there, it was just defined twice in `lms/envs/aws.py`. I removed the second one, leaving the first: https://github.com/open-craft/edx-platform/blob/6aad123fe57710d9ee23aeae62506f2808abd087/lms/envs/aws.py#L136

bradenmacdonald commented:
@omarkhan 

I tried configuring two providers, one using http://placehold.it/60x60 and one using http://placehold.it/100x200 . 

Both images get stretched to be square (so one is being distorted), and they still don't have enough spacing. They need to match the size of the FA icons.
![wrong](https://cloud.githubusercontent.com/assets/945577/13993493/e1afb396-f0dd-11e5-96b8-60daf116a844.png)

Here's how it should look:
![correct](https://cloud.githubusercontent.com/assets/945577/13993531/fe9eabd8-f0dd-11e5-8ac9-c8cc6cb125a1.png)

That can be achieved with these rules for the `img`:

```css
img {
    width: auto;
    height: auto;
    max-height: 1.4em;
    margin-top: -2px;
    max-width: 1.4em;
}
```

bradenmacdonald commented:
@omarkhan When I try to update an existing `oauth2providerconfig` that has an image already, and I don't want to change the image, I get this error message: "Either an icon class or an icon image must be given (but not both)"

i.e. I leave the image setting alone, so it looks like this:
![screen shot 2016-03-23 at 10 10 23 am](https://cloud.githubusercontent.com/assets/945577/13993895/82953866-f0df-11e5-8b5c-27c95c80a787.png)

Then I press save and get this:
![screen shot 2016-03-23 at 10 11 05 am](https://cloud.githubusercontent.com/assets/945577/13993910/990e571c-f0df-11e5-96c5-0049b3a48cd4.png)
(Notice the existing image disappeared after I pressed Save)

bradenmacdonald commented:
@omarkhan The button hover color also needs to match the logo background color:
![screen shot 2016-03-23 at 10 17 38 am](https://cloud.githubusercontent.com/assets/945577/13994133/9cb8f420-f0e0-11e5-86bb-b456808bdb81.png)

(It does match for the standard providers, e.g. Google is red when you hover over it).

bradenmacdonald commented:
Hinted login is easy to test - if you have the Dummy OAuth2 provider configured, just go to http://localhost:8000/courses/course-v1:edX+DemoX+Demo_Course/courseware/?tpa_hint=oa2-dummy

You can see it's currently quite broken:
![screen shot 2016-03-23 at 10 26 03 am](https://cloud.githubusercontent.com/assets/945577/13994364/aabb76e6-f0e1-11e5-9e6e-c259d812fbb8.png)

bradenmacdonald commented:
Also, please remove the screenshots of the old registration form from the PR description since we don't need to support that.

omarkhan commented:
Thanks for the review @bradenmacdonald. I will push an update shortly

omarkhan commented:
@bradenmacdonald I have pushed the changes you requested. I haven't been able to fix the django admin issue yet - it looks like there is some custom admin logic in there, and the edit form isn't actually editing the instance but adding a new one. I'm not sure how this works, can you clarify?

The sandbox is reprovisioning....

bradenmacdonald commented:
@omarkhan Yep, this is a keyed configuration model, which means it saves a complete history of all changes. Every time you edit, it creates a new object, and hides the older objects in the admin (though you can show them using a filter on the right).

The logic for creating a new entry is here: https://github.com/edx/edx-platform/blob/406066ca51ac9be85b9c8f6a461636c0fde90c78/common/djangoapps/config_models/admin.py#L177-L182

I think you'll need to subclass `KeyedConfigurationModelAdmin` and add in the correct behavior for copying the filename of the existing image to the new instance. Or override the validation used when saving the model. I'm not sure of what the best approach there is.

omarkhan commented:
@bradenmacdonald I'll look into that tomorrow morning. The sandbox is up: http://pr11871.sandbox.opencraft.com/login

For some reason font awesome is not working. I'm having the same problem in my devstack, but it works in the bok choy tests so I suspect it has something to do with the asset pipeline. i just started another ansible run on the sandbox, let's see if that sorts it out.

bradenmacdonald commented:
@omarkhan Would you like to ask @pomegranited to take a look at fixing that image configuration issue while you sleep?

omarkhan commented:
@pomegranited your fix works when you leave the icon image unchanged, but it fails when you try to clear or change the image:

- When you change the image, it just resets it to a copy of the previous image
- When you clear the image, you get this:
<img width="412" alt="screen shot 2016-03-24 at 16 55 00" src="https://cloud.githubusercontent.com/assets/793379/14013929/e0ab2958-f1e1-11e5-86d6-0f79a67be124.png">

pomegranited commented:
@omarkhan Ooh sorry, thank you for catching that.  Fixed with [04aece3](https://github.com/edx/edx-platform/pull/11871/commits/04aece39d1a3dbafbd5079be784cac9fe221c699)

omarkhan commented:
Thanks Jill!

bradenmacdonald commented on common/djangoapps/config_models/admin.py:
```diff
@@ -178,8 +182,38 @@ def add_view(self, request, form_url='', extra_context=None):
+                clear_checkbox = False
+                if form_field and hasattr(form_field, 'widget') and hasattr(form_field.widget, 'clear_checkbox_name'):
+                    clear_checkbox = CheckboxInput().value_from_datadict(
+                        request.POST, request.FILES, form_field.widget.clear_checkbox_name(field_name))
```
nit: put closing parent onto next line:

```python
clear_checkbox = CheckboxInput().value_from_datadict(
    request.POST, request.FILES, form_field.widget.clear_checkbox_name(field_name)
)
```

bradenmacdonald commented on common/djangoapps/config_models/admin.py:
```diff
@@ -178,8 +182,38 @@ def add_view(self, request, form_url='', extra_context=None):
+                        file_content = default_storage.open(field_value)
+                        request.FILES[field_name] = SimpleUploadedFile.from_dict({
+                            'filename': unicode(field_value),
+                            'content': file_content.read()
```
Not a big fan of this approach... If the file is hosted on S3, this approach would require that the LMS download the file and upload it again, which wastes time and bandwidth in order to accomplish nothing.

Is there not some way that we can just say, "if the user has not chosen to clear or replace the file, set the string value of the file path stored in the database on the new model to be the same as the string value on the source model."

i.e. In the database, django model FileFields are simply represented as a string path, e.g. "some-logo.png". All we really need to do is to bypass the file validation and uploading logic to set the raw value of the new field to the same string as the source model, i.e. "some-logo.png".

omarkhan replied:
OK I'll try that.

omarkhan replied:
Fixed in d6540a8

bradenmacdonald commented:
@omarkhan Very sorry but one more wrinkle: I noticed you've changed the text color to match the icon background color, which is great, but from our mockup for this project you'll notice we need to support the background color `#00BCF2`. And if we use that color for text on a white background, [it doesn't have enough contrast](http://contrastchecker.com/). So we need to either add an optional "override text color" field, or use [this](http://stackoverflow.com/a/3943023/1057326) to detect when the custom color's contrast is too low, and use the standard (dark blueish) color in that case.

bradenmacdonald commented:
@jzoldak or @andy-armstrong , would one of you be able to take a quick look at the bok choy tweaks made in this WIP PR, and let us know if they're OK? @omarkhan had to tweak MEDIA_URL to get file uploads to work properly. Specifically, I'm referring to the changes in `cms/envs/bok_choy.env.json`, `lms/envs/bok_choy.env.json`, `lms/envs/bok_choy.py`, and `pavelib/utils/test/suites/bokchoy_suite.py`.

CC @benpatterson

bradenmacdonald commented:
@omarkhan d6540a8 looks like a nicer approach, but now when I go to "Update" an entry, the Image field is not set, and "Clear" checkbox is missing. So it correctly preserves the image, but there's no way to remove the image, and it looks like there's no image set even when there is.

omarkhan commented:
@bradenmacdonald you can still clear the image, it's just a bit of a 2-step process:

- Set an `icon_class` value, save
- The form complains about setting both an image and an icon class. This time the clear checkbox should appear. Check it and save again

This is not ideal, but coming up with a better solution here is non-trivial.

omarkhan commented:
@bradenmacdonald I have added a separate `text_color` field. It will use the `button_color` value as a fallback.

omarkhan commented:
@bradenmacdonald looks like the admin fix was trivial after all: 1fd9d05

I have updated the sandbox, will follow up in the morning

bradenmacdonald commented:
@omarkhan Nice, thanks! Just tested and it seems to work well.

bradenmacdonald commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -415,10 +435,6 @@ $sm-btn-linkedin: #0077b5;
-
-            &:hover {
-                box-shadow: 0 2px 1px 0 #8D3024;
-            }
```
What's the reason for removing these? So you didn't have to add `box-shadow-color: ...` to the inline styles? I think it's actually a nice change but it won't be my call.

omarkhan replied:
That was for consistency. "Normal" SSO buttons (not google, facebook, etc)
don't have a box-shadow on hover.

bradenmacdonald replied:
Hmm, on http://sso.sandbox.opencraft.com/login they do... But it is inconsistent. I personally think this is better.

bradenmacdonald commented:
Thanks for the push on this, @omarkhan. LGTM! :+1: if tests pass.

omarkhan commented:
Commits squashed for upstream review

benpatterson commented:
@bradenmacdonald @omarkhan I had a bit of time to look this morning, and I had some basic questions. 
* Will the icon be hashed & properly cached? (Easier to ask you then try to grok the code for an answer)
* I don't think the `MEDIA_URL` change is an issue for bok-choy. However, I'm not sure why you would need to copy something from common over to the test_root. Why is that?

benpatterson commented on pavelib/utils/test/suites/bokchoy_suite.py:
```diff
@@ -188,6 +190,13 @@ def load_data(self):
+        Copies upload fixtures into MEDIA_ROOT
+        """
+        print 'Copying fixtures from uploads directory to MEDIA_ROOT'
+        sh('cp -rf common/test/uploads/* test_root/uploads/')
```
Should: Rather than doing this via shell, you should do it via python. 
Question: I also have a broader question on why this should be necessary. 
Nit: I think copying the whole directory is the kind of thing that becomes more expensive as the codebase grows. Would it make sense to have a stricter scope (e.g., only the file you care about?)
Should: I notice that you're not actually leveraging the environment `MEDIA_ROOT` setting being used by bok-choy here, but instead it is hardcoded. Should you reference the var? (I realize this may not be possible since it's a django setting, but I wanted to ask if you have any more-robust recommendations.) Again this makes me wonder why you'd want to copy assets anyway, rather than storing them in both places <gasp!>

omarkhan replied:
The reason for copying these files over here is that `test_root/uploads/` is the `MEDIA_ROOT` for the bokchoy tests, and pretty much everything that ends up there is included in the `.gitignore`. So I figured it would make sense to put the fixtures somewhere else (`common/test/` is where the other fixtures live) and copy them over when the tests are run.

I can copy them over using python rather than the shell if you prefer.

I would prefer to reference the `MEDIA_ROOT` setting here, but we don't have access to it yet as we haven't loaded the django settings.

We could avoid this altogether by loading these fixtures somewhere in the test setup, rather than by using `./manage.py loaddata`.

benpatterson replied:
Ok I see what you mean. Did you happen to look at whether or not `MEDIA_ROOT` can be set to something else, like a subdir? I'm not too familiar with this setting but wanted to ask.

To your point about avoiding this altogether, I think if we had the right directory to begin with, then we could avoid any copying. What other options are there?

If we must copy over from another directory, I'd rather the copy be more explicit, and I think it should be done via python. But, I want to hear about our other options first because I feel like copying the data over can be brittle.

omarkhan replied:
@benpatterson I have avoided the copying by just putting the fixture directly in `test_root/uploads/`, as per your suggestion.

bradenmacdonald commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -356,6 +356,22 @@ $sm-btn-linkedin: #0077b5;
+            height: auto;
+            max-height: 1.4em;
+            max-width: 1.4em;
+            margin-top: -2px;
```
I just changed the sandbox to use the logo we need for this project, and it seems to be not vertically centered unless this margin-top is -4px. @omarkhan can you please check?

The logo as SVG is attached to OC-1374 if you want to use it for devstack testing.

omarkhan replied:
You're right, -4px seems to do it. Done in aa5dd88

bradenmacdonald commented on common/djangoapps/third_party_auth/tests/test_admin.py:
```diff
@@ -0,0 +1,71 @@
+    """
+    Tests for oauth2 provider config admin
+    """
+    def test_oauth2_provider_edit_icon_image(self):
```
Please add a docstring that clearly explains why this test is important, given our use of a KeyedConfigurationModel with a FileField.

omarkhan replied:
Done in 43254c2

omarkhan commented:
@benpatterson there is no provision here for image cacheing. How these images will be served will depend on the storage setup - filesystem, S3... Hashing and cache headers can be configured at the storage layer. 

For example, the default file system storage will rename files with a unique hash if they have the same filename. Throw in some nginx configuration and you have client-side image cacheing.

For S3, I believe you can configure the cache headers for a bucket using the aws command line tool.

AlasdairSwan commented on lms/templates/student_account/register.underscore:
```diff
@@ -29,11 +29,17 @@
+                                    <img src="<%- provider.iconImage %>" alt="<%- provider.name %> icon" />
+                                <% } %>
+                            </div>
+                            <span <% if ( provider.textColor ) { %>style="color: <%- provider.textColor %>"<% } %> aria-hidden="true"><%- provider.name %></span>
```
Adding the text colour as an inline value here does not allow for updating the hover state colour. I don't think there is any need to use inline styles for styling the button and its text It would be better to assign the button a custom class and style it via Sass/CSS so you can override the various button states while continuing to use stylesheets for all styling.

omarkhan replied:
Does that mean that you will have to include a custom theme to change the button color? CC @bradenmacdonald

AlasdairSwan replied:
You could just add a third-party-auth partial that could be updated per instance to set the styles they want.

bradenmacdonald replied:
@AlasdairSwan We need to use this on edx.org among other instances. Is there a mechanism for adding a custom SASS partial to an instance (ultimately via ansible/configuration), other than via a custom theme?

If not, I suppose we could add custom classes for the AzureAD provider to the global sass files where the styles for Google, Facebook, and LinkedIn already reside, and say that for other custom providers, instances must use a custom theme.

AlasdairSwan replied:
@bradenmacdonald I spoke with @douglashall earlier this morning and either he or @mattdrayer will be able to provide support re. how to include this in Sass. Apologies but I am not very familiar with theming.

mattdrayer replied:
It will probably be most efficient to have @saleem-latif jump in here with some advice re: Sass partials and theming.

douglashall replied:
I'm not sure theming should come into play here at all. Is there a requirement that we customize these styles based on the provider?

bradenmacdonald replied:
Users can add arbitrary/custom providers\*, so they need to have *some* mechanism to specify the color/style and icon of their custom providers. Our current approach is just that they specify the icon and colors to use when configuring the provider, but @AlasdairSwan is suggesting we find a way to use sass to control the appearance of any custom buttons. As far as I know, that would require each instance that wants to use custom providers to style their buttons within a theme, right?

\* e.g. Imagine I want to add GitHub login on my instance, or SAML login to my local university.

bradenmacdonald replied:
@AlasdairSwan @douglashall @omarkhan 
I need to de-risk this ASAP so that we can make this Friday's release. Our main focus for this project is adding the AzureAD provider, but we want to do so in a generic way to avoid introducing tech debt. So: Are there any objections to this plan?
* we add classes for the AzureAD provider to the global sass files where the styles for Google, Facebook, and LinkedIn already reside
* We remove "background color" and "text color" from the django admin
* We remove the inline styles from this PR, which would no longer be needed

With the above changes, we wouldn't provide an explicit mechanism for styling other custom providers, but it could be done by adding custom sass to any instance, either through a theme or some other mechanism.

AlasdairSwan replied:
@bradenmacdonald I am fine with adding the styles to the Sass file where the other provider styles are defined.

mattdrayer replied:
Yes, this sounds like a good approach to me as well.  Let's leave the door open for @saleem-latif to respond re: custom sass via theming, because it might end up being something you can do quickly/easily -- however no need to block on that so I would plan to move ahead with what you are suggesting.

douglashall replied:
@bradenmacdonald Sounds good to me.

omarkhan replied:
@bradenmacdonald @AlasdairSwan @douglashall @mattdrayer I have made these changes, see my latest commits

AlasdairSwan commented on themes/stanford-style/lms/templates/register-form.html:
```diff
@@ -26,7 +26,14 @@ <h3 class="message-title">${_("The following errors occurred while processing yo
+        % if enabled.icon_class:
+        <span class="icon fa ${enabled.icon_class}" aria-hidden="true"></span>
+        % else:
+        <img class="icon" src="${enabled.icon_image.url}" alt="${enabled.name} icon" aria-hidden="true" />
```
If the img is aria-hidden there is no value in adding alt text. A cleaner way to achieve the same thing would be to set ```alt=""``` which will hide the image from screenreaders.

AlasdairSwan commented on lms/static/sass/multicourse/_account.scss:
```diff
@@ -526,6 +526,14 @@
         color: inherit;
         @include margin-right($baseline/2);
+
+        img {
```
Classes should be used instead of element selectors.

AlasdairSwan commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -356,6 +356,22 @@ $sm-btn-linkedin: #0077b5;
+            max-height: 1.4em;
+            max-width: 1.4em;
+            margin-top: -4px;
+        }
```
This selector duplicates the values set in _account.scss. If you use a class you could reuser the styles with the duplication.

AlasdairSwan commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -378,13 +394,21 @@ $sm-btn-linkedin: #0077b5;
+                max-height: 1.4em;
+                max-width: 1.4em;
+                margin-top: -4px;
+            }
```
This selector duplicates the values set in _account.scss. If you use a class you could reuser the styles with the duplication.

AlasdairSwan commented on lms/templates/student_account/register.underscore:
```diff
@@ -29,11 +29,17 @@
-                        <span aria-hidden="true"><%- provider.name %></span>
-                        <span class="sr"><%- _.sprintf( gettext("Create account using %(providerName)s."), {providerName: provider.name} ) %></span>
-                    </button>
+                    <div class="button-container" <% if ( provider.buttonColor ) { %>style="color: <%- provider.buttonColor %>; background-color: <%- provider.buttonColor %>"<% } %>>
```
I still think a provider specific class should be added here and the styling handled by updating a third party auth Sass partial instead of using inline styles. It will make the code much more flexible and enable full editing of button states, and it is best practice to avoid inline styles unless absolutely necessary.

benpatterson commented:
@omarkhan Ok I think I understand your point re: caching. When the image is uploaded, it follows the same paradigm as course assets that are uploaded?

omarkhan commented:
@benpatterson I don't know how course assets work. This just defers to the default django storage backend, configured with the `DEFAULT_FILE_STORAGE` setting.

saleem-latif commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -6,6 +6,7 @@
 $sm-btn-google: #dd4b39;
 $sm-btn-facebook: #3b5998;
 $sm-btn-linkedin: #0077b5;
+$sm-btn-azuread: #00bcf2;
```
@omarkhan it would be better if these definitions could go in `partials/base/variables` inside lms/static it would make it easy for custom themes to override these variables. would it be feasible to move these definitions to `partials/base/_variables.scss` in this PR ?

omarkhan replied:
Sure, I will make this change now

saleem-latif commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -464,6 +465,25 @@ $sm-btn-linkedin: #0077b5;
+            &:focus {
+                background-color: $sm-btn-azuread;
+                border: 1px solid $sm-btn-azuread;
+                color: white;
```
@omarkhan I think it would be better to use variable `$white` instead of `white`.

omarkhan replied:
OK. Any particular reason or is it just convention?

saleem-latif replied:
just a convention to make it consistent overall.

omarkhan commented:
@saleem-latif I have implemented the changes you requested

benpatterson commented on test_root/uploads/.gitignore:
```diff
@@ -1,4 +0,0 @@
-*.csv
```
Instead of removing this file, you could keep it, but add an exception for the files you want to be able to checkin. Something like this:
```bash
*.csv
*.jpg
*.png
*.txt
!my_file.jpg
```

omarkhan replied:
@benpatterson done

benpatterson commented:
In terms of the items I needed to review (e.g., bok-choy configuration, etc), things LGTM. Thanks @omarkhan

bradenmacdonald commented:
Thanks for the review @benpatterson .

@AlasdairSwan please take another look as soon as you're able.

AlasdairSwan commented:
Looks good @bradenmacdonald :+1:

marcotuts commented:
very thorough documentation and clear communication on this. Can't say how awesome it is to also have a sandbox to preview this kind of change. Thanks everyone! :+1:

bradenmacdonald commented:
@omarkhan Please squash, check that tests pass, then merge.

omarkhan commented:
Commits squashed

bradenmacdonald commented:
@edx/devops Heads up that this PR is targeting Friday's release and contains a migration. I will also need to work with one of you to configure and test the new provider next week - will discuss tomorrow.
